src/hotspot/share/opto/loopTransform.cpp

```diff
@@ -3112,6 +3132,35 @@ void PhaseIdealLoop::do_range_check(IdealLoopTree *loop, Node_List &old_new) {
   // The OpaqueNode is unshared by design
   assert(opqzm->outcnt() == 1, "cannot hack shared node");
   _igvn.replace_input_of(opqzm, 1, main_limit);
+  // new main_limit can push Bool/Cmp nodes down (when one of the eliminated condition has parameters that are not loop
+  // invariant in the pre loop.
+  set_ctrl(opqzm, new_limit_ctrl);
+  set_ctrl(iffm->in(1), new_limit_ctrl);
+  set_ctrl(iffm->in(1)->in(1), new_limit_ctrl);
+}
+
+// Adjust control for node and its inputs (and inputs of its inputs) to be above the pre end
+void PhaseIdealLoop::ensure_node_and_inputs_are_above_pre_end(CountedLoopEndNode* pre_end, Node* node, Node*& control) {
```

Comment:
`control` should be the `ctrl` of `node` here, right (i.e. `control == get_ctrl(node)`)? Maybe we can assert that. But thinking about it, would it hurt to just call `get_ctrl(node)` again here and remove `control` as parameter? Then we can just return the new control and do 
```
offset_c = ensure_node_and_inputs_are_above_pre_end(pre_end, offset);
```
at the call-site.